### PR TITLE
✨Source Xero: Migrate Python CDK to Low-code CDK

### LINKED ISSUE #36866: Source Xero: Migrate Python CDK to Low-code CDK
### Topic

Migrate Xero to low code

### Relevant information

https://docs.airbyte.com/integrations/sources/xero

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
*Migrating Source Xero to Low-Code CDK*
Closes airbytehq/airbyte-internal-issues#7187

## How
*Developed using (Configuration Based Source) low-code CDK*

## Recommended reading order
1. `spec.yaml`
2. `manifest.yaml`
3. `schemas/*`

## Tests

<details open><summary><strong>Integration & Acceptance</strong></summary>

 - Test successful with full refresh

Screenshot from 2024-04-07 00-28-47

</details>

## 🚨 User Impact 🚨
- Migration to low-code

### CODE REVIEW

**Comment by btkcodedev** [comment]
Need help with incremental sync, currently, the TestTwoSequentialReads Test is failing,
The backend is not changing the slice as per timestamp I think,
Comments by reviewers would be very helpful
image

File: airbyte-integrations/connectors/source-xero/source_xero/components.py
Comment (girarda) [inline airbyte-integrations/connectors/source-xero/source_xero/components.py:110]: can you document why this cannot be a standard oauth authenticator?

Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/components.py:110]: It is a SingleUseRefreshToken and after checking /connections endpoint to verify the 200 status, it might need to get new token via client_id:client_secret logic
image

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:1]: This is extremely outdated. Let's make sure that you re-import connector into Builder with running custom components locally. Happy to show how to run them.

File: airbyte-integrations/connectors/source-xero/source_xero/components.py
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-xero/source_xero/components.py:38]: This looks suspicious. Sure, I get it, it works, but I wonder if we can just replace that with a format matcher instead in the low-code declaration?

Comment (natikgadzhi) [reply airbyte-integrations/connectors/source-xero/source_xero/components.py:38]: I.e. I know that people run Xero connector in Cloud Builder — and they deal with dates somehow. They don't have access to custom components whatsoever. We could grab an example I guess 👀

**Review by natikgadzhi — CHANGES_REQUESTED** [review]
Change of plans. Let's fully remove OAuth support from this source, and leave it to only work with API token auth. 

The reason is — for now — our Xero OAuth app hit the quota and we're trying to partner up with them, but for now — we want to release the connector to cloud and folks to be able to use it, but for that we need to remove OAuth as an option.

This means two things: 
- Removing OAuth and using standard token authenticataor (yay, one less custom component!)
- This is a breaking change, so breaking change needs to be documented accordingly.

File: airbyte-integrations/connectors/source-xero/source_xero/components.py
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/components.py:38]: But it destructs the backward compatibility of those fields, instead of DateTime it would provide `"/Date(1630412754013+0000)/"`

Also, the incremental sync is built with one of these as cursor field
- "UpdatedDateUTC": "/Date(1630412754013+0000)/",

we need to convert each of those fields to standard datetime format,
Else on second thought, If incremental is removed, we only need the connector builder and no external strings needed.

Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/components.py:38]: The response is just giving a timestamp with Date() wrapped around, 
- "UpdatedDateUTC": "/Date(1630412754013+0000)/",

CustomExtractor searches every nestings of response and converts it to standard date-time
If we remove it, It will change the schema and expected records and probably destination properties for anyone using it,

I've updated the Auth scheme to bearer, And every test is passing locally,
Is there any other way other than a custom component? It has to search every nested loop
Please comment if there are any possibilities.

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml]: Updated and tested with connector builder :white_check_mark:, 
Apart from custom extractor, if it's DpathExtractor, There are no problems, we could use the manifest directly to generate UI in builder
image

**Comment by btkcodedev** [comment]
Reviewers, please change the access_token in the cloud for running the CI

**Comment by lazebnyi** [comment]
@btkcodedev please write your requests in slack chat to speed up process 
and can you specify their  how we need change `access_token` ?

**Review by darynaishchenko — CHANGES_REQUESTED** [review]
@btkcodedev, Could you please return unit tests and change  their implementation to use declarative streams? Don't need to test low-code related logic, only specific cases for streams.  Also we should add unit tests for custom components.

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:43]: Could you please set page to 100 as it was in previous version, or there are some reasons why we changing it?

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:58]: Why we don't use `If-Modified-Since` header logic for incremental sync as it was before and filter data instead? It can be a big regression of sync time for users with a lot of data.

File: airbyte-integrations/connectors/source-xero/source_xero/components.py
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/source_xero/components.py:106]: can you please divide this method into class methods for more readability?

File: airbyte-integrations/connectors/source-xero/acceptance-test-config.yml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/acceptance-test-config.yml:51]: Can you add more light about why we skip sequential read? It was added to fix your issue described here https://github.com/airbytehq/airbyte/pull/36956#issuecomment-2046706792? Probably it occurred due to second read with state from the first read returned same records as the first read, which we don't expect. Can you reproduce this behavior locally?

File: docs/integrations/sources/xero-migrations.md
Comment (darynaishchenko) [inline docs/integrations/sources/xero-migrations.md:6]: Please add steps that users need to perform if they used OAuth? authenticate the source with access token etc

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:43]: Done

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:58]: There is a fault in the API precision as it only supports date precision right now, we need a second precision for incremental reads. Mentioned in  https://github.com/airbytehq/airbyte/pull/36956/files#diff-ad6ad0ce85c924cf3ece039ff15425fd7c28cd5df20d06343b6fe8a5bd67b799R6

File: airbyte-integrations/connectors/source-xero/acceptance-test-config.yml
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/acceptance-test-config.yml:51]: Yes, I could reproduce this behavior locally,
With the help of client-side Incremental reads, TestTwoSequentialReads is passed
However, TestReadSequentialSlices is skipped due to this behavior.

File: docs/integrations/sources/xero-migrations.md
Comment (btkcodedev) [reply docs/integrations/sources/xero-migrations.md:19]: Done

File: airbyte-integrations/connectors/source-xero/source_xero/components.py
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-xero/source_xero/components.py:111]: Done

**Comment by btkcodedev** [comment]
Hi @darynaishchenko, All comments are resolved and brought back the unit_tests, All passing in local, the access_token is expired in the cloud,
It has a very short duration, please ping me @btkcodedev on slack for getting freshly prepared creds during next review

**Comment by btkcodedev** [comment]
image

File: docs/integrations/sources/xero-migrations.md
Comment (darynaishchenko) [inline docs/integrations/sources/xero-migrations.md:17]:
```suggestion
Click `Get New Access Token` for retrieving access token

Then authorize your source with `access_token`.
1. Go to set up `The Source` page.
2. Enter your Xero application's access token.
3. Click `Reset saved source` button. 

```

**Comment by btkcodedev** [comment]
CI all green, All comments resolved, Please continue merge process :bow:

**Comment by darynaishchenko** [comment]
> CI all green, All comments resolved, Please continue merge process 🙇

@btkcodedev I'm going to run regression tests on your pr. Let you know the results. Thanks!

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:29]: I guess we should add 401 status code also with error message asking user to update access token and failed action.

File: airbyte-integrations/connectors/source-xero/metadata.yaml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/metadata.yaml:19]:
```suggestion
                    The source Xero connector is being migrated from the Python CDK to our declarative low-code CDK. The authorization method was changed from OAuth 2.0 to Access Token. Due to changes in the authorization users will need to generate their Access Token and reset source configuration. For more information, see our migration documentation for source Xero.
```

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:472]: Please change it to `type object ` for every stream schema, because actual declaration causes passing `$parameters ` into schema.
```suggestion
    type: object
```

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:30]:
```suggestion
              error_message:  Failed to authorize request. Please update your access token to continue using the source.
```

**Review by darynaishchenko — APPROVED** [review]

File: airbyte-integrations/connectors/source-xero/metadata.yaml
Comment (katmarkham) [inline airbyte-integrations/connectors/source-xero/metadata.yaml:19]:
```suggestion
          The source Xero connector is being migrated from the Python CDK to our declarative low-code CDK. The authorization method was changed from OAuth 2.0 to Access Token. Due to changes in authorization, users will need to generate an access token within source Xero and reset source configuration. For more information, see our migration documentation for source Xero.
```

File: docs/integrations/sources/xero-migrations.md
Comment (katmarkham) [inline docs/integrations/sources/xero-migrations.md:5]:
```suggestion
The authentication schema is now using `access_token` instead of Oauth 2.0. Visit the Xero documentation - https://developer.xero.com/documentation/guides/oauth2/pkce-flow for more detailed information about how to get access token.
```

File: docs/integrations/sources/xero-migrations.md
Comment (katmarkham) [inline docs/integrations/sources/xero-migrations.md:6]:
```suggestion
Optionally, you may get your access_token via Postman:
```

File: docs/integrations/sources/xero-migrations.md
Comment (katmarkham) [inline docs/integrations/sources/xero-migrations.md:25]:
```suggestion
As Xero API now only supports date-precision, instead of second precision filtering through If-Modified-Since header, reads are now streamlined to incremental through client side.
```

**Review by katmarkham — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-xero/source_xero/manifest.yaml:1]: The version here is old, let's aim for 1.0.0

File: airbyte-integrations/connectors/source-xero/pyproject.toml
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-xero/pyproject.toml:19]: Pin to `^1`

**Review by natikgadzhi — APPROVED** [review]
Rubberstamping to unblock. 
With caveats of getting @katmarkham changes in, and updating CDK version to 1.0 — this is fine. 

You will need to resolve conflicts though ;(

**Review by katmarkham — APPROVED** [review]

**Comment by ak-workit** [comment]
Thank u team!!

**Pull request merged**
